azurerm/internal/services/storage/resource_arm_hpc_cache.go

```diff
@@ -0,0 +1,229 @@
+func resourceArmHPCCache() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmHPCCacheCreate,
+		Read:   resourceArmHPCCacheRead,
+		Delete: resourceArmHPCCacheDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"location": azure.SchemaLocation(),
+
+			"cache_size": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.IntInSlice([]int{
+					3072,
+					6144,
+					12288,
+					24576,
+					49152,
+				}),
+			},
+
+			"subnet_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: azure.ValidateResourceIDOrEmpty,
+			},
+
+			"sku_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"Standard_2G",
+					"Standard_4G",
+					"Standard_8G",
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceArmHPCCacheCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Storage.CachesClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for Azure HPC Cache creation.")
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
+		existing, err := client.Get(ctx, resourceGroup, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing HPC Cache %q (Resource Group %q): %s", name, resourceGroup, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_hpc_cache", *existing.ID)
+		}
+	}
+
+	location := d.Get("location").(string)
+	cacheSize := d.Get("cache_size").(int)
+	subnet := d.Get("subnet_id").(string)
+	skuName := d.Get("sku_name").(string)
+
+	cache := &storagecache.Cache{
+		Name:     utils.String(name),
+		Location: utils.String(location),
+		CacheProperties: &storagecache.CacheProperties{
+			CacheSizeGB: utils.Int32(int32(cacheSize)),
+			Subnet:      utils.String(subnet),
+		},
+		Sku: &storagecache.CacheSku{
+			Name: utils.String(skuName),
+		},
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, resourceGroup, name, cache); err != nil {
+		return fmt.Errorf("Error creating HPC Cache %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	stateConf := &resource.StateChangeConf{
```

Comment:
@aqche 

I may be dumb here, why not just use `future.WaitForCompletionRef()` here? Is there some limitation about this API?

Reply:
@magodo thanks for calling that out, it actually slipped my mind to try `future.WaitForCompletionRef()` here first. I'll give it a test and report back.